### PR TITLE
Enables pure IPv6 configurations for nd_setup module on ND version 3.0.1 and later (DCNE-241)

### LINKED ISSUE #106: Support for IPv6-only ND Setup
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

My customer is building a new SW-defined data center using IPv6 only. IPv6-only is mandatory for compliance with the IPv6 Master Plan for the German federal public administrations, as well as various EU regulations. This includes all management networks and tools.

Beginning with release 3.0(1), Nexus Dashboard supports pure IPv6 configurations for the cluster nodes and services. However in the current version of the nd_setup module, various IPv4 parameters are required. For IPv6-only support, IPv4 and IPv6 parameters need to become choices instead. For example, ipv4_address/ipv4_gateway only, ipv6_address/ipv6_gateway only, or both. Similar, app_network/service_network and app_network_ipv6/service_network_ipv6.

### Example

```
  - name: Set up ND
    cisco.nd.nd_setup:
      cluster_name: vND-Cluster-3
      ntp_config:
        servers:
          - ntp_host: 2a00:8a60:0:20::123
      dns_server:
        - 2001:420:464d:1::a
      app_network_ipv6: 2000::/108
      service_network_ipv6: 3000::/108
      deployment_mode:
        - ndfc
      external_services:
        management_service_ips:
          - fd00:16:1:ac::35
          - fd00:16:1:ac::36
          - fd00:16:1:ac::37
      nodes:
        - hostname: vND3-node1
          serial_number: 6BAEC7FB40CD
          username: rescue-user
          password: xxxxx
          management_network:
            ipv6_address: fd00:16:1:ac::30/64
            ipv6_gateway: fd00:16:1:ac::1
          data_network:
            ipv6_address: fd00:16:1:ad::30/64
            ipv6_gateway: fd00:16:1:ad::1
```

### New or Affected Module(s)

* nd_setup

### ND Version and ND Platform

* Version: 3.0(1), 3.1.1, 3.2.1, or higher
* Platform: all

### Collection Versions

* cisco.nd 1.2.0, 1.3.0

### PR DESCRIPTION
**Pull request by gmicol**
fixes #106

### CODE REVIEW

**Comment by akinross** [comment]
black and sanity issues

File: plugins/modules/nd_setup.py [hunk 1]
```diff
@@ -442,6 +447,14 @@ def main():
...
         nd.existing = nd.request("/clusterstatus/install", method="GET")
     else:
+        nd_version = nd.query_obj("/version.json")
+        nd_version = ".".join(str(nd_version[key]) for key in ["major", "minor", "maintenance"])
+        if nd_version >= "3.0.1":
+            if not all((app_network, service_network)) and not all((app_network_ipv6, service_network_ipv6)):
+                nd.fail_json(msg="Application and service network addresses, IPv4 or IPv6, are required during ND setup.")
+        else:
+            if not all((app_network, service_network)):
+                nd.fail_json(msg="Application and service network IPv4 addresses are required during ND setup.")
```
Comment (shrsr) [inline plugins/modules/nd_setup.py:457]: Shall we move check_network_requirements() outside the payload initialization for consistency? We usually validate attributes before setting the payload? The older payload will remain as is which is more readable?

```
for node in nodes:
def check_network_requirements(node, nd_version):
        network = node.get('management_network', {})
        if nd_version >= "3.0.1":
            if not all(network.get(ip) for ip in ["ipv6Subnet", "gatewayv6"]) and not all(network.get(ip) for ip in ["ipSubnet", "gateway"]):
                nd.fail_json(msg="Missing IPv4 subnet/gateway or IPv6 subnet/gateway in node {0} configuration.".format(type))
        else:
            if not all(network.get(ip) for ip in ["ipSubnet", "gateway"]):
                nd.fail_json(msg="Missing IPv4 subnet/gateway in node {0} configuration.".format(type))
```

**Review by shrsr — CHANGES_REQUESTED** [review]

File: plugins/modules/nd_setup.py (hunk 1 above)
Comment (gmicol) [reply plugins/modules/nd_setup.py]: I moved it back before payload initialization as you suggested.

**Review by shrsr — APPROVED** [review]
LGTM

**Review by samiib — CHANGES_REQUESTED** [review]

File: tests/integration/targets/nd_setup/tasks/main.yml [hunk 2]
```diff
@@ -265,6 +265,24 @@
...
+      <<: *nd_setup_test_config
+      nodes:
+        - hostname: Test
+          serial_number: "{{ serial_number }}"
+          role: primary
+          management_ip_address: "{{ management_ip_address }}"
+          username: "{{ deployment_username | default('rescue-user') }}"
+          password: "{{ deployment_password }}"
+          management_network:
+            ipv4_gateway: "{{ management_gateway }}"
```
[the remark this answers is not in the document]
Comment (gmicol) [reply tests/integration/targets/nd_setup/tasks/main.yml:280]: Yes, we don't have any ND cluster using IPv6. However, you're right about adding negative test for IP requirements. I will add them. thanks

**Review by akinross — APPROVED** [review]
LGTM

**Review by samiib — APPROVED** [review]
LGTM

**Review by shrsr — APPROVED** [review]
LGTM

**Review by lhercot — APPROVED** [review]
LGTM

**Pull request merged**